compiler/resolution/fixupExports.cpp

```diff
@@ -0,0 +1,425 @@
+// interop.
+//
+static const char* getConversionCallName(Type* srctype, Type* dsttype) {
+  const char* result = NULL;
+
+  // TODO: We should probably maintain this in a table or the like.
+  if (srctype == dtString && dsttype == dtStringC) {
+    result = "chpl__exportStringToConstCharPtr";
+  } else if (srctype == dtStringC && dsttype == dtString) {
+    result = "chpl__exportConstCharPtrToString";
+  } else if (srctype == dtString && dsttype == getCharPtrType()) {
+    result = "chpl__exportStringToCharPtr";
+  } else if (srctype == getCharPtrType() && dsttype == dtString) {
+    result = "chpl__exportCharPtrToString";
+  } else {
+    INT_FATAL("Bad types for conversion call in: %s", __FUNCTION__);
+  }
+
+  return result;
+}
+
+static FnSymbol* getConversionCall(Type* srctype, Type* dsttype) {
+  const char* name = getConversionCallName(srctype, dsttype);
+  INT_ASSERT(name != NULL);
+
+  std::map<const char*, FnSymbol*>& m = conversionCallMap;
+  std::map<const char*, FnSymbol*>::iterator it;
+  it = m.find(name);
+  if (it != m.end()) { return it->second; }
+
+  FnSymbol* result = resolveConversionCall(srctype, dsttype);
+  m[name] = result;
+
+  return result;
+}
+
+//
+// TODO: This assumes that the names of the conversion calls are unique and
+// _are not_ being overloaded.
+//
+static FnSymbol* resolveConversionCall(Type* srctype, Type* dsttype) {
+  const char* name = getConversionCallName(srctype, dsttype);
+
+  BlockStmt* block = new BlockStmt();
+  VarSymbol* tmp = newTemp(srctype);
+  CallExpr* call = new CallExpr(name, tmp);
+  FnSymbol* result = NULL;
+ 
+  // Modifications to `chpl_gen_main` are only temporary.
+  chpl_gen_main->body->insertAtHead(block);
+
+  block->insertAtTail(new DefExpr(tmp));
+  block->insertAtTail(call);
+
+  tryResolveCall(call);
+
+  result = call->resolvedFunction();
+  INT_ASSERT(result != NULL);
+
+  //
+  // These are not called by user code paths, so we must resolve them now or
+  // we will get AST errors during late stage checking.
+  //
+  normalize(result);
+  resolveFunction(result);
+
+  call->remove();
+  tmp->defPoint->remove();
+  block->remove();
+
+  return result;
+}
+
+static VarSymbol* fixupFormal(FnSymbol* wrapper, int idx) {
+  ArgSymbol* as = wrapper->getFormal(idx);
+  Type* otype = as->type;
+  Type* wtype = getTypeForFixup(otype);
+
+  as->type = wtype;
+
+  //
+  // Now create a new temporary, and assign it the result of a conversion call
+  // on the formal we are performing a fixup on.
+  //
+
+  VarSymbol* tmp = newTemp(otype);
+  VarSymbol* result = tmp;
+
+  wrapper->body->insertAtTail(new DefExpr(tmp));
+
+  FnSymbol* conv = getConversionCall(wtype, otype);
+  INT_ASSERT(conv != NULL);
+  
+  CallExpr* call = new CallExpr(conv, as);
+  CallExpr* move = new CallExpr(PRIM_MOVE, tmp, call);
+  wrapper->body->insertAtTail(move);
+  
+  return result;
+}
+
+static void changeRetType(FnSymbol* wrapper) {
+  Type* otype = wrapper->retType;
+  INT_ASSERT(needsFixup(otype));
```

Comment:
I don't think this is a useful INT_ASSERT given that the callers of this function check it first.

Reply:
Yeah, I would agree that I have a lot of redundant asserts.